src/auth.ts

```diff
@@ -0,0 +1,81 @@
+/*
+auth.ts - Provides auth related methods for the Bluebutton class
+*/
+import crypto from "crypto";
+
+import BlueButton from ".";
+import { authData, tokenPostData } from "./types/auth";
+
+function base64URLEncode(buffer: Buffer): string {
+  return buffer
+    .toString("base64")
+    .replace(/\+/g, "-")
+    .replace(/\//g, "_")
+    .replace(/=/g, "");
+}
+
+function sha256(str: string): Buffer {
+  return crypto.createHash("sha256").update(str).digest();
+}
+
+type pkceData = {
+  codeChallenge: string;
+  verifier: string;
+};
+
+function generatePkceData(): pkceData {
+  var verifier = base64URLEncode(crypto.randomBytes(32));
+  return {
+    codeChallenge: base64URLEncode(sha256(verifier)),
+    verifier: verifier,
+  };
+}
+
+function generateRandomState(): string {
+  return base64URLEncode(crypto.randomBytes(32));
+}
+
+export function generateAuthData(): authData {
+  const pkceData = generatePkceData();
+  return {
+    codeChallenge: pkceData.codeChallenge,
+    verifier: pkceData.verifier,
+    state: generateRandomState(),
+  };
+}
+
+export function generateAuthorizeUrl(
+  bb: BlueButton,
+  authData: authData
+): string {
+  const BB2_AUTH_URL = bb.baseUrl + "/" + bb.version + "/o/authorize";
+
+  const pkceParams = `code_challenge_method=S256&code_challenge=${authData.codeChallenge}`;
+
+  return `${BB2_AUTH_URL}?client_id=${bb.clientId}&redirect_uri=${bb.callBackUrl}&state=${authData.state}&response_type=code&${pkceParams}`;
+}
+
+//  Generates post data for call to access token URL
+export function generateTokenPostData(
```

Comment:
Hmmm I'm questioning the value add of this function... we might just be able to remove it? I'm also a bit worried about returning sensitive data in the response. Might make it easier for users to accidentally log out something sensitive unintentionally.

Reply:
@nbragdon Was initially thinking we wouldn't be making actual requests via axios in the SDK, so was providing this. Just about to work on the `getAccessToken()` method, so will make this function private to it.

Reply:
@nbragdon I removed `generateTokenPostData()` and `validateCallbackRequestQueryParams()` from the `BlueButton` class methods.

They are utilized inside of the `getAuthorizationToken()` method now.